OpenRA.Mods.Common/Traits/Armament.cs

```diff
@@ -141,15 +147,20 @@ public Armament(Actor self, ArmamentInfo info)
 
 		public virtual WDist MaxRange()
 		{
-			return new WDist(Util.ApplyPercentageModifiers(Weapon.Range.Length, rangeModifiers));
+			return new WDist(Util.ApplyPercentageModifiers(Weapon.Range.Length, rangeModifiers.Select(m => m.GetRangeModifier())));
 		}
 
 		protected override void Created(Actor self)
 		{
 			turret = self.TraitsImplementing<Turreted>().FirstOrDefault(t => t.Name == Info.Turret);
 			ammoPool = self.TraitsImplementing<AmmoPool>().FirstOrDefault(la => la.Info.Name == Info.AmmoPoolName);
 			coords = self.Trait<BodyOrientation>();
-			rangeModifiers = self.TraitsImplementing<IRangeModifier>().ToArray().Select(m => m.GetRangeModifier());
+			notifyBurstComplete = self.TraitsImplementing<INotifyBurstComplete>().ToArray();
+			reloadModifiers = self.TraitsImplementing<IReloadModifier>().ToArray();
+			rangeModifiers = self.TraitsImplementing<IRangeModifier>().ToArray();
+			damageModifiers = self.TraitsImplementing<IFirepowerModifier>().ToArray();
+			inaccuracyModifiers = self.TraitsImplementing<IInaccuracyModifier>().ToArray();
+			notifyAttacks = self.TraitsImplementing<INotifyAttack>().ToArray();
```

Comment:
Remember that LINQ execution is deferred. All we create is a query, but the results of that query are not gathered until we actually enumerate the sequence.

```c#
var numbers = new List<int> { 1, 2, 3 };
var multipler = 2;
var multipliedNumbers = numbers.Select(n => n * multipler);

var doubled = string.Join(", ", multipliedNumbers); // "2, 4, 6"
multipler = 3; // Let's change the results - without having to Select again.
var tripled = string.Join(", ", multipliedNumbers); // "3, 6, 9"
```

Notice I create the query with `.Select` just once, but the expression `n => n * multipler` is evaluated fresh each time - which means I can change just the `multiplier` and generate new numbers.